### PR TITLE
[OpenMP][OMPIRBuilder] Refactor reduction initialization logic into one util

### PR DESCRIPTION
**Pull request by ergawy**
This refactors the logic needed to emit init logic for reductions by moving some duplicated code into a shared util. The logic for doing is quite involved and is needed for any construct that has reductions. Moreover, when a construct has both private and reduction clauses, both sets of clauses need to cooperate with each other when emitting the logic needed for allocation and initialization. Therefore, this PR clearly sets the boundaries for the logic needed to initialize reductions.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp [hunk 1]
```diff
@@ -1929,6 +1974,7 @@ convertOmpParallel(omp::ParallelOp opInst, llvm::IRBuilderBase &builder,
           splitBB(builder, /*CreateBranch=*/true, "omp.private.copy");
       builder.SetInsertPoint(copyBlock->getFirstNonPHIOrDbgOrAlloca());
     }
+
```
Comment (luporl) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:1977]: Unrelated change?

**Review by luporl — COMMENTED** [review]
LGTM, but someone more familiar with this code should approve.

**Review by tblah — APPROVED** [review]
Thanks for the cleanup!

**Comment by tblah** [comment]
This breaks some of my test cases. Here's a reproducer:
```
Program test
  real :: x(1)
  integer :: i
  !$OMP PARALLEL DO REDUCTION(+:x)
    Do i = 1,1
      x = 1
    End Do
  !$OMP END PARALLEL DO
End Program
```

Do you have time to look at this (in the new year is fine) or would you like me to take a look? Apologies that it took me so long to get around to running my downstream tests.

**Comment by ergawy** [comment]
Thanks for the heads-up @tblah. Looking into this now...
